### PR TITLE
Transforms for dew point temperature and air density

### PR DESCRIPTION
**Pull request by GHBLoos**
Dew Point Temperature added for BME 280 sensor.

### CODE REVIEW

**Comment by mairas** [comment]
Hey, kudos for your first PR! Dew point calculation is definitely a good feature to have. Since it is a function of temperature and humidity (and pressure), it would also be usable for other hardware sensors (and architecturally more apt) if it was a transform instead of part of the BME280 sensor.

Also, while at it, shouldn't it be using the pressure-aware formulations such as the Arden Buck equation? https://en.wikipedia.org/wiki/Dew_point#Calculating_the_dew_point

**Comment by GHBLoos** [comment]
I can improve the code using the Arden Buck equation. But as I read it, you can use humidity _or_ pressure for the equation. So you don't need both.
A transform may be a good solution. Would it be possible to use temperature and humidity (or pressure) from two different sensors? I am just not good enough in coding to figure that out myself.

File: src/transforms/dew_point.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#include "dew_point.h"
+
+// dew point
+
+DewPoint::DewPoint() : NumericTransform() {}
+
+void DewPoint::set_input(float input, uint8_t inputChannel) {
+  inputs[inputChannel] = input;
+  received |= 1 << inputChannel;
+  if (received == 0b11) {
```
Comment (ba58smith) [inline src/transforms/dew_point.cpp:10]: I'd like to see a comment briefly explaining what `if (received == 0b11)` is doing.

File: src/transforms/dew_point.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    received = 0;
+
+          float TC= inputs[0]-273.15;     // temperature in Celsius
+          float RH= inputs[1];            // relative humidity
+
```
Comment (ba58smith) [inline src/transforms/dew_point.cpp:15]: spacing, for consistency and readability. As previously noted, we've spent a lot of time making the SensESP code follow the Google Style Guide, and be consistent.

above, in the example: (bme_temperature, bme_humidity)

TC = inputs
RH = inputs
and below:
TC < 0
gamma = log(RH * exp((b - (TC / d)) * (TC / (c + TC))));
(c * gamma) / (b - gamma);

**Review by ba58smith — APPROVED** [review]
Nice job! Just a few housekeeping comments.

File: src/transforms/dew_point.cpp (hunk 1 above)
Comment (GHBLoos) [reply src/transforms/dew_point.cpp:10]: I copied this from difference.cpp. To be honest, I don't really understand why it is like that.

File: src/sensors/bme280.cpp [hunk 3]
```diff
@@ -46,7 +47,20 @@ void BME280Value::enable() {
...
+
+          float h= bme280_->adafruit_bme280_->readHumidity()/100; //humidity in percent, so divide by 100
+          float t= bme280_->adafruit_bme280_->readTemperature();
+
+          float alfa=(a*t)/(b+t)+ log(h);
+
+          float dpt = (b*alfa)/(a-alfa); 
+        
+          output = dpt + 273.15; // Kelvin is Celsius + 273.15
+       } else {
```
[the remark this answers is not in the document]
Comment (GHBLoos) [reply src/sensors/bme280.cpp:68]: Ok, working on that. Thanks for pointing me to the Google Style Guide. That helps.

[the remark this answers is not in the document]
Comment (GHBLoos) [reply src/sensors/bme280.cpp:68]: Looking through that guide:
`enum BME280ValType { temperature, pressure, humidity, dewPointTemperature };`
Shouldn't that be like
`enum BME280ValType { kTemperature, kPressure, kHumidity, kDewPointTemperature };`
(https://google.github.io/styleguide/cppguide.html#Enumerator_Names)

[the remark this answers is not in the document]
Comment (ba58smith) [reply src/sensors/bme280.cpp:73]: You're correct. We don't follow the Guide to the letter,  but we do try to be very consistent with the parts we do follow. If you want to use the leading k in this Transform, I don't think anyone will object,  and maybe you'll start a trend!

[the remark this answers is not in the document]
Comment (mairas) [reply src/sensors/bme280.cpp]: Regarding code formatting: if you're using Visual Studio Code you can get your code automatically formatted (spaces, indentation, line widths, whatnot) according to the project standards by applying the command "Format Document" (either via Cmd-P or Ctrl-P, or by shortcut Alt-Shift-F - not sure if the latter is like that on every platform, though).

[the remark this answers is not in the document]
Comment (GHBLoos) [reply src/sensors/bme280.cpp]: Learned something again today! Thanks. That saves a lot of work.

**Comment by GHBLoos** [comment]
Sorry for the mess.

**Comment by GHBLoos** [comment]
I think this is completed now. I am not sure how to finish this process. Do I have to do anything? Or are you going to decide to merge it or not. Let me know if I need to do something.

File: examples/bme280_example.cpp [hunk 4]
```diff
@@ -52,5 +54,22 @@ ReactESP app([]() {
...
+
+  // Use the transform dewPoint to calculate the dewpoint based upon the temperature and humidity.
+  auto* dew_point  = 
+      new DewPoint();
+
+  dew_point->connect_from(bme_temperature, bme_humidity)
+          ->connect_to(new SKOutputNumber("environment.outside.dewPointTemperature"));
+
+  // Use the transform airDensity to calculate the air density of humid air ased
+  // upon the temperature, humidity and pressure.
```
Comment (ba58smith) [inline examples/bme280_example.cpp:66]: Typo: "ased" should be "based".

File: examples/bme280_example.cpp [hunk 5]
```diff
@@ -52,5 +54,22 @@ ReactESP app([]() {
 
   bme_humidity->connect_to(new SKOutputNumber("environment.outside.humidity"));
 
+
+  // Use the transform dewPoint to calculate the dewpoint based upon the temperature and humidity.
+  auto* dew_point  = 
+      new DewPoint();
```
Comment (ba58smith) [inline examples/bme280_example.cpp:60]: Did the PlatformIO auto-format break the above into two lines? Looks weird. Of course, it works either way, but seems like it should be a single line.

File: examples/bme280_example.cpp [hunk 6]
```diff
@@ -52,5 +54,22 @@ ReactESP app([]() {
...
+
+  dew_point->connect_from(bme_temperature, bme_humidity)
+          ->connect_to(new SKOutputNumber("environment.outside.dewPointTemperature"));
+
+  // Use the transform airDensity to calculate the air density of humid air ased
+  // upon the temperature, humidity and pressure.
+  auto* airDensity = 
+      new AirDensity();
```
Comment (ba58smith) [inline examples/bme280_example.cpp:68]: Same comment as above about this being two lines instead of one.

**Review by ba58smith — APPROVED** [review]
One typo, two very minor formatting questions. Other than that, it looks good to me. Nice job!

I would like @mairas to do the final review and merge.

File: examples/bme280_example.cpp (hunk 5 above)
Comment (GHBLoos) [reply examples/bme280_example.cpp]: I am not sure, but I put it on a single line now.

**Comment by mairas** [comment]
This looks great! Good work! Git commit history could be reworked a bit but I don't have the heart to torture you with it so I'll just squash and merge. :-)

**Pull request merged**